src/Pact/Types/Type.hs

```diff
@@ -60,55 +61,46 @@ import Pact.Types.Info
 
 
 newtype TypeName = TypeName Text
-  deriving (Eq,Ord,IsString,AsString,ToJSON,FromJSON,Pretty,Generic,NFData)
-instance Show TypeName where show (TypeName s) = show s
+  deriving (Eq,Ord,IsString,AsString,ToJSON,FromJSON,Pretty,Generic,NFData,Show)
 
 -- | Pair a name and a type (arguments, bindings etc)
 data Arg o = Arg {
   _aName :: Text,
   _aType :: Type o,
   _aInfo :: Info
-  } deriving (Eq,Ord,Functor,Foldable,Traversable,Generic)
+  } deriving (Eq,Ord,Functor,Foldable,Traversable,Generic,Show)
 
 instance NFData o => NFData (Arg o)
-instance Show o => Show (Arg o) where show (Arg n t _) = unpack n ++ ":" ++ show t
 instance (Pretty o) => Pretty (Arg o)
-  where pretty (Arg n t _) = pretty n PP.<> colon PP.<> pretty t
-
-instance Eq1 Arg where
-  liftEq eq (Arg a b c) (Arg m n o) = a == m && liftEq eq b n && c == o
+  where pretty (Arg n t _) = pretty n <> colon <> pretty t
 
 -- | Function type
 data FunType o = FunType {
   _ftArgs :: [Arg o],
   _ftReturn :: Type o
-  } deriving (Eq,Ord,Functor,Foldable,Traversable,Generic)
+  } deriving (Eq,Ord,Functor,Foldable,Traversable,Generic,Show)
 
 instance NFData o => NFData (FunType o)
-instance Show o => Show (FunType o) where
-  show (FunType as t) = "(" ++ unwords (map show as) ++ " -> " ++ show t ++ ")"
 instance (Pretty o) => Pretty (FunType o) where
-  pretty (FunType as t) = parens (hsep (map pretty as) <+> "->" <+> pretty t)
-
-instance Eq1 FunType where
-  liftEq eq (FunType a b) (FunType m n) = liftEq (liftEq eq) a m && liftEq eq b n
+  pretty (FunType as t) = hsep (map pretty as) <+> "->" <+> pretty t
```

Comment:
I agree with @emilypi  that these are pretty bad, but the problem is how to represent a funtype in a lispy way. The current is both hard to read and non-lispy:

```
x:<a [string....] y:<a [foo....] -> string
```

Unfortunately the literature is not inspiring, c.f. <https://docs.racket-lang.org/ts-guide/types.html#(part._.Function_.Types)> which imports `->` from Haskell presumably ...

In Pact source, defuns have the following syntax:

```
(defun foo:string (bar:integer baz:object{quux}))
```

This is why I went with the half-assed format above, moving the return type to the right.

Perhaps `(foo:string bar:integer):object{quux}`? This might be an acceptable user format for lambda/partial args if we ever get there.

Reply:
I'm going to punt on this for this review